src/v/rpc/service.h

```diff
@@ -78,10 +78,10 @@ struct service::execution_helper {
     static ss::future<netbuf> exec(
       ss::input_stream<char>& in,
       streaming_context& ctx,
-      uint32_t method_id,
+      method_info method,
```

Comment:
since method_info is `static constexpr` can we accept const reference here ?

Reply:
See the above discussion regarding cheap-to-pass. Generally for cheap-to-pass values, the recommentation is to use pass-by-value, though it is not an absolute: if you are forwarding a cheap-to-pass struct that takes more than 8 bytes (as here) through many functions, pass-by-reference can be cheaper as it needs to move only 1 pointer to the right register, versus at least two values making up the stucture otherwise.

See `pass_value_to_value_offset` vs `pass_ref_to_ref_offset` in [this example](https://godbolt.org/z/sM6Toh4WY), and note that we need the "offset" to ensure that the passed parameter is not the same position in the argument list in the caller and callee since otherwise there is no shuffling needed in either case.

However, when you actually go to use the value, the pass-by-value function will likely be cheaper since the values are in registers. Additionally, the compiler will have to reload the values from memory less frequently or not at all since by-value results in a private, not-escaped copy in each function, whereas pass-by-reference points to values which may be modified by almost any write to memory, call to unknown function, etc.

Reply:
Other note, about `static constexpr` specifically: passing by reference will actually result in the need to materialize the constant `method_info` object in the binary so that its address can be taken. [Example](https://godbolt.org/z/xxrGdGe1n):

```
method1_by_value():                  # @method1_by_value()
        mov     edi, offset method1
        jmp     takes_by_ref(method_info const&) # TAILCALL
method1:
        .quad   .L.str
        .long   1                               # 0x1
        .zero   4

.L.str:
        .asciz  "method1"
```

`method1` is the constexpr object in the `.rodata` section, which appears in the final binary.

On the other hand, pass by value can use the constexpr object directly and if the result is pass-in-register, no object is created in the binary at all: as [seen here](https://godbolt.org/z/WEYq7r5zP) where the `method1` object has disappeared.

Reply:
this is cool travis. i've often wondered about some of these details.

Reply:
>passing by reference will actually result in the need to materialize the constant method_info object in the binary so that its address can be taken

!

Reply:
this is great, thank you very much Travis